irohad/ordering/impl/on_demand_ordering_service_impl.cpp

```diff
@@ -5,24 +5,135 @@
 
 #include "ordering/impl/on_demand_ordering_service_impl.hpp"
 
+#include "builders/protobuf/proposal.hpp"
+#include "datetime/time.hpp"
+#include "interfaces/iroha_internal/proposal.hpp"
+#include "interfaces/transaction.hpp"
+
 using namespace iroha::ordering;
 
-// ------------------ | OnDemandOrderingService |-------------------------
+OnDemandOrderingServiceImpl::OnDemandOrderingServiceImpl(
+    size_t transaction_limit,
+    size_t number_of_proposals,
+    const transport::RoundType &initial_round)
+    : transaction_limit_(transaction_limit),
+      number_of_proposals_(number_of_proposals),
+      current_proposal_(std::make_pair(
+          initial_round, tbb::concurrent_queue<TransactionType>())),
+      log_(logger::log("OnDemandOrderingServiceImpl")) {}
+
+// -------------------------| OnDemandOrderingService |-------------------------
 
 void OnDemandOrderingServiceImpl::onCollaborationOutcome(
     RoundOutput outcome, transport::RoundType round) {
-  throw std::logic_error("Function not yet implemented");
+  log_->info(
+      "onCollaborationOutcome => round[{}, {}]", round.first, round.second);
+  // exclusive write lock
+  std::lock_guard<std::shared_timed_mutex> guard(lock_);
+  log_->info("onCollaborationOutcome => write lock is acquired");
+
+  packNextProposal(outcome, round);
+  tryErase();
 }
 
-// -------------------- | OdOsNotification |------------------------------
+// ----------------------------| OdOsNotification |-----------------------------
 
 void OnDemandOrderingServiceImpl::onTransactions(
-    std::vector<std::shared_ptr<shared_model::interface::Transaction>>
-        transactions) {
-  throw std::logic_error("Function not yet implemented");
+    const CollectionType &transactions) {
+  // read lock
+  std::shared_lock<std::shared_timed_mutex> guard(lock_);
+  log_->info("onTransactions => collections size = {}", transactions.size());
+
+  std::for_each(
```

Comment:
Maybe use insert or similar function instead of for_each?

Reply:
What are you imagine?